README.md
# dart_jsona

It's a ported to dart - [Jsona TS](https://github.com/olosegres/jsona)

Framework agnostic, customizable library that provide data formatters to simplify work with JSON API v1.0 specification.

This package don't have any dependencies and pretty simple. It used `Map<String, dynamic>` a lot 😑

It was designed to work with [json_serializable](https://pub.dartlang.org/packages/json_serializable)

## Usage

### Deserializer - creates simplified object(s) from json

```
import 'package:dart_jsona/dart_jsona.dart';

main() {
  Jsona jsona = new Jsona();

  final body = {
    'data': {
      'type': 'town',
      'id': '123',
      'attributes': {
        'name': 'Barcelona',
      },
      'relationships': {
        'country': {
          'data': {
            'type': 'country',
            'id': '32',
          }
        }
      }
    },
    'included': [{
      'type': 'country',
      'id': '32',
      'attributes': {
        'name': 'Spain',
      }
    }]
  };

  final town = jsona.deserialize(body);

  print(town);
  /**
    {
      type: town,
      id: 123,
      name: Barcelona,
      country: {
        type: country,
        id: 32, name: Spain
      },
      relationshipNames: [country]
    }
   */
}
```

### Serializer - creates json from simplified object(s)
```
import 'package:dart_jsona/dart_jsona.dart';
main() {
    Jsona jsona = new Jsona();

    Map<String dynamic> newJson = jsona.serialize(stuff: town, includedNames: ['country']);

    print(newJson);

    /**
       {
         data: {
           id: 123,
           type: town,
           attributes: {
             name: Barcelona
           },
           relationships: {
             country: {
               data: {
                 id: 32, type: country
               }
             }
           }
         },
         included: [{
           id: 32,
           type: country,
           attributes: {
             name: Spain
           }
         }]
       }
    */
}
```

### Deserialize example from https://jsonapi.org/

```
var jsonApi = {
  "links": {
    "self": "http://example.com/articles",
    "next": "http://example.com/articles?page[offset]=2",
    "last": "http://example.com/articles?page[offset]=10"
  },
  "data": [{
    "type": "articles",
    "id": "1",
    "attributes": {
      "title": "JSON API paints my bikeshed!"
    },
    "relationships": {
      "author": {
        "links": {
          "self": "http://example.com/articles/1/relationships/author",
          "related": "http://example.com/articles/1/author"
        },
        "data": { "type": "people", "id": "9" }
      },
      "comments": {
        "links": {
          "self": "http://example.com/articles/1/relationships/comments",
          "related": "http://example.com/articles/1/comments"
        },
        "data": [
          { "type": "comments", "id": "5" },
          { "type": "comments", "id": "12" }
        ]
      }
    },
    "links": {
      "self": "http://example.com/articles/1"
    }
  }],
  "included": [{
    "type": "people",
    "id": "9",
    "attributes": {
      "firstName": "Dan",
      "lastName": "Gebhardt",
      "twitter": "dgeb"
    },
    "links": {
      "self": "http://example.com/people/9"
    }
  }, {
    "type": "comments",
    "id": "5",
    "attributes": {
      "body": "First!"
    },
    "relationships": {
      "author": {
        "data": { "type": "people", "id": "2" }
      }
    },
    "links": {
      "self": "http://example.com/comments/5"
    }
  }, {
    "type": "comments",
    "id": "12",
    "attributes": {
      "body": "I like XML better"
    },
    "relationships": {
      "author": {
        "data": { "type": "people", "id": "9" }
      }
    },
    "links": {
      "self": "http://example.com/comments/12"
    }
  }]
};

var jsonApiModel = jsona.deserialize(jsonApi);

print(jsonApiModel);

/**
  [
    {
      type: 'articles',
      id: 1,
      title: 'JSON API paints my bikeshed!',
      links: {
        self: 'http://example.com/articles/1'
      },
      author: {
        type: 'people',
        id: 9,
        firstName: 'Dan',
        lastName: 'Gebhardt',
        twitter: 'dgeb',
        links: {
          self: 'http://example.com/people/9'
        }
      },
      comments: [{
        type: 'comments',
        id: 5,
        body: 'First!',
        links: {
          self: 'http://example.com/comments/5'
        },
        author: {
          type: 'people',
          id: 2
        },
        relationshipNames: [author]
      },
      {
        type: 'comments',
        id: 12,
        body: 'I like XML better',
        links: {
          self: 'http://example.com/comments/12'
        },
        author: {
          type: 'people',
          id: 9,
          firstName: 'Dan',
          lastName: 'Gebhardt',
          twitter: 'dgeb',
          links: {
            self: 'http://example.com/people/9'
          }
        },
        relationshipNames: [author]
      }],
      relationshipNames: [author, comments]
    }
  ];
**/
```

### Deserialize in angular service + `dart:convert` + [json_serializable](https://pub.dartlang.org/packages/json_serializable)

```
// post_service.dart

import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart';
import 'package:dart_jsona/dart_jsona.dart';

class PostService {

  static final _postUrl = 'api/posts';
  final Client _http;

  PostService(this._http);

  Future<List<PostModel>> getPosts() async {
    try {
      final response = await _http.get('${apiUrl}/${_postUrl}');

      // create a new post from json
      final posts = (_extractData(response) as List)
          .map((value) => PostModel.fromJson(value))
          .toList();
      return posts;
    } catch(e) {
      throw _handleError(e);
    }
  }

  dynamic _extractData(Response resp) {
    // initialze Jsona
    Jsona jsona = new Jsona();

    // convert to `Map<String, dynamic>` from `String`
    var response = json.decode(resp.body)

    // deserialize response body
    return jsona.deserialize(response);
  }

  Exception _handleError(dynamic e) {
    return Exception('Server error; cause: $e');
  }
}
```

```
// post_model.dart

import 'package:json_annotation/json_annotation.dart';

part 'post_model.g.dart';

@JsonSerializable()
class PostModel {
  final String body;
  final String image;
  PostModel(this.body, this.image);

  factory PostModel.fromJson(Map<String, dynamic> json) => _$PostModelFromJson(json);
  Map<String, dynamic> toJson() => _$PostModelToJson(this);
}
```

## Contributing

Contributing are welcome.

This package on an early stage, and I'll be happy if you fix or create a new feature.

You just need follow these steps:
- Fork the repo
- Make new branch
- Write code magic
- Write tests
- Create PR

## License
Jsona, examples provided in this repository and in the documentation are [MIT licensed](./LICENSE).


CHANGELOG.md
## 0.0.1
- Create Deserializer and Simple Serializer
```
Jsona jsona = new Jsona();
final town = jsona.deserialize(body);

final newJson = jsona.serialize(stuff: town, includedNames: ['country']);
```


test/mock.dart
import 'package:dart_jsona/src/builders/property_mapper.dart' show RELATIONSHIP_NAMES_PROP;

final Map<String, dynamic> country1 = {
  'model': {
    'type': 'country',
    'id': '34',
    'name': 'Spain'
  },
  'json': {
    'type': 'country',
    'id': '34',
    'attributes': {
      'name': 'Spain'
    }
  }
};

final Map<String, dynamic> country2 = {
  'model': {
    'type': 'country',
    'id': '25',
    'name': 'China'
  },
  'json': {
    'type': 'country',
    'id': '25',
    'attributes': {
      'name': 'China'
    }
  }
};

final Map<String, dynamic> town1 = {
  'model': {
    'type': 'town',
    'id': '21',
    'name': 'Shanghai',
    'country': country1['model'],
    RELATIONSHIP_NAMES_PROP: ['country']
  },
  'json': {
    'type': 'town',
    'id': '21',
    'attributes': {
      'name': 'Shanghai',
    },
    'relationships': {
      'country': {
        'data': {
          'type': 'country',
          'id': country1['model']['id'],
        }
      }
    }
  }
};


final Map<String, dynamic> town2 = {
  'model': {
    'type': 'town',
    'id': '80',
    'name': 'Barcelona',
    'country': country2['model'],
    RELATIONSHIP_NAMES_PROP: ['country']
  },
  'json': {
    'type': 'town',
    'id': '80',
    'attributes': {
      'name': 'Barcelona',
    },
    'relationships': {
      'country': {
        'data': {
          'type': 'country',
          'id': country2['model']['id'],
        }
      }
    }
  }
};

final Map<String, dynamic> specialty1 = {
  'model': {
    'type': 'specialty',
    'id': '1',
    'title': 'mycategory1'
  },
  'json': {
    'type': 'specialty',
    'id': '1',
    'attributes': {
      'title': 'mycategory1'
    }
  }
};


final Map<String, dynamic> specialty2 = {
  'model': {
    'type': 'specialty',
    'id': '2',
    'title': 'mycategory2'
  },
  'json': {
    'type': 'specialty',
    'id': '2',
    'attributes': {
      'title': 'mycategory2'
    }
  }
};

final Map<String, dynamic> user1 = {
  'model': {
    'type': 'user',
    'id': '1',
    'name': 'myName1',
    'active': false,
    'town': town1['model'],
    'specialty': [specialty1['model']],
    RELATIONSHIP_NAMES_PROP: ['town', 'specialty']
  },
  'json': {
    'type': 'user',
    'id': '1',
    'attributes': {
      'name': 'myName1',
      'active': false,
    },
    'relationships': {
      'town': {
        'data': {
          'type': 'town',
          'id': town1['model']['id'],
        }
      },
      'specialty': {
        'data': [{
          'type': 'specialty',
          'id': specialty1['model']['id']
        }]
      }
    }
  },
  'included': {
    'townOnly': [
      town1['json']
    ]
  }
};

final Map<String, dynamic> user2 = {
  'model': {
    'type': 'user',
    'id': '2',
    'name': 'myName2',
    'active': true,
    'town': town2['model'],
    'specialty': [specialty1['model'], specialty2['model']],
    RELATIONSHIP_NAMES_PROP: ['town', 'specialty']
  },
  'modelWithoutIncluded': {
    'type': 'user',
    'id': '2',
    'name': 'myName2',
    'active': true,
    'town': {
      'id': town2['model']['id'],
      'type': town2['model']['type'],
    },
    'specialty': [{
      'id': specialty1['model']['id'],
      'type': specialty1['model']['type'],
    }, {
      'id': specialty2['model']['id'],
      'type': specialty2['model']['type'],
    }],
    RELATIONSHIP_NAMES_PROP: ['town', 'specialty']
  },
  'json': {
    'type': 'user',
    'id': '2',
    'attributes': {
      'name': 'myName2',
      'active': true,
    },
    'relationships': {
      'town': {
        'data': {
          'type': 'town',
          'id': town2['model']['id'],
        }
      },
      'specialty': {
        'data': [{
          'type': 'specialty',
          'id': specialty1['model']['id']
        }, {
          'type': 'specialty',
          'id': specialty2['model']['id']
        }]
      }
    }
  },
  'includeNames': {
    'townOnly': ['town']
  },
  'included': {
    'townOnly': [
      town2['json']
    ]
  }
};

final article1 = {
  'model': {
    'type': 'article',
    'id': 1,
    'likes': 5550,
    'author': user1['model'],
    'country': country1['model'],
    RELATIONSHIP_NAMES_PROP: ['author', 'country']
  },
  'json': {
    'type': 'article',
    'id': 1,
    'attributes': {
      'likes': 5550
    },
    'relationships': {
      'author': {
        'data': {
          'type': 'user',
          'id': user1['model']['id']
        }
      },
      'country': {
        'data': {
          'type': 'country',
          'id': country1['model']['id']
        }
      }
    }
  }
};

final article2 = {
  'model': {
    'type': 'article',
    'id': 2,
    'likes': 100,
    'author': user2['model'],
    'country': country2['model'],
    RELATIONSHIP_NAMES_PROP: ['author', 'country']
  },
  'json': {
    'type': 'article',
    'id': 2,
    'attributes': {
      'likes': 100
    },
    'relationships': {
      'author': {
        'data': {
          'type': 'user',
          'id': user2['model']['id'],
        }
      },
      'country': {
        'data': {
          'type': 'country',
          'id': country2['model']['id'],
        }
      }
    }
  },
  'includeNames': [
    'author.town.contry',
    'author.specialty',
    'country'
  ],
};

final articleWithoutAuthor = {
  'model': {
    'type': 'article',
    'id': 3,
    'likes': 0,
    'author': null,
    RELATIONSHIP_NAMES_PROP: ['author']
  },
  'json': {
    'type': 'article',
    'id': 3,
    'attributes': {
      'likes': 0,
    },
    'relationships': {
      'author': {
        'data': null,
      },
    },
  },
  'includeNames': [
    'author'
  ],
};


final jsonApiExample = {
  "links": {
    "self": "http://example.com/articles",
    "next": "http://example.com/articles?page[offset]=2",
    "last": "http://example.com/articles?page[offset]=10"
  },
  "data": [{
    "type": "articles",
    "id": "1",
    "attributes": {
      "title": "JSON API paints my bikeshed!"
    },
    "relationships": {
      "author": {
        "links": {
          "self": "http://example.com/articles/1/relationships/author",
          "related": "http://example.com/articles/1/author"
        },
        "data": { "type": "people", "id": "9" }
      },
      "comments": {
        "links": {
          "self": "http://example.com/articles/1/relationships/comments",
          "related": "http://example.com/articles/1/comments"
        },
        "data": [
          { "type": "comments", "id": "5" },
          { "type": "comments", "id": "12" }
        ]
      }
    },
    "links": {
      "self": "http://example.com/articles/1"
    }
  }],
  "included": [{
    "type": "people",
    "id": "9",
    "attributes": {
      "firstName": "Dan",
      "lastName": "Gebhardt",
      "twitter": "dgeb"
    },
    "links": {
      "self": "http://example.com/people/9"
    }
  }, {
    "type": "comments",
    "id": "5",
    "attributes": {
      "body": "First!"
    },
    "relationships": {
      "author": {
        "data": { "type": "people", "id": "2" }
      }
    },
    "links": {
      "self": "http://example.com/comments/5"
    }
  }, {
    "type": "comments",
    "id": "12",
    "attributes": {
      "body": "I like XML better"
    },
    "relationships": {
      "author": {
        "data": { "type": "people", "id": "9" }
      }
    },
    "links": {
      "self": "http://example.com/comments/12"
    }
  }]
};

final jsonApiModel = [
  {
    'type': 'articles',
    'id': 1,
    'title': 'JSON API paints my bikeshed!',
    'links': {
      'self': 'http://example.com/articles/1'
    },
    'author': {
      'type': 'people',
      'id': 9,
      'firstName': 'Dan',
      'lastName': 'Gebhardt',
      'twitter': 'dgeb',
      'links': {
        'self': 'http://example.com/people/9'
      }
    },
    'comments': [{
      'type': 'comments',
      'id': 5,
      'body': 'First!',
      'links': {
        'self': 'http://example.com/comments/5'
      },
      'author': {
        'type': 'people',
        'id': 2
      },
      'relationshipNames': ['author']
    },
    {
      'type': 'comments',
      'id': 12,
      'body': 'I like XML better',
      'links': {
        'self': 'http://example.com/comments/12'
      },
      'author': {
        'type': 'people',
        'id': 9,
        'firstName': 'Dan',
        'lastName': 'Gebhardt',
        'twitter': 'dgeb',
        'links': {
          'self': 'http://example.com/people/9'
        }
      },
      'relationshipNames': ['author']
    }],
    'relationshipNames': ['author', 'comments']
  }
];


test/dart_jsona_test.dart
import 'package:test/test.dart';

import 'package:dart_jsona/dart_jsona.dart';

import 'mock.dart';

void main() {
  group('Jsona', () {
    Jsona jsona;

    setUp(() {
      jsona = new Jsona();
    });

    group('deserialize', () {
      test('should deserialize item without included', () {
        var userModel = jsona.deserialize({'data': user2['json']});

        expect(userModel, equals(user2['modelWithoutIncluded']));
      });

      test('should desserialize collection without included', () {
        var townsCollection = jsona.deserialize({
          'data': [town1['json'], town2['json']],
          'included': [country1['json'], country2['json']]
        });

        expect(townsCollection, equals([town1['model'], town2['model']]));
      });

      test('should deserialize example from JSONapi.org', () {

        dynamic jsonApiModel = jsona.deserialize(jsonApiExample);

        expect(jsonApiModel, equals(jsonApiModel));
      });
    });

    group('serialize', () {
      test('should build json with item, without included', () {
        var jsonBody = jsona.serialize(stuff: town1['model']);

        expect(jsonBody['data'], equals(town1['json']));
        expect(jsonBody['included'], equals(null));
      });

      test('should build json with collection, with included', () {
        var jsonBody = jsona.serialize(stuff: user2['model'], includeNames: ['specialty', 'town']);

        expect(jsonBody['data'], equals(user2['json']));
        expect(jsonBody['included'], equals([specialty1['json'], specialty2['json'], town2['json']]));
      });
    });
  });
}


test/test_classes/person.dart
import 'package:json_annotation/json_annotation.dart';

part 'person.g.dart';

@JsonSerializable()
class Person {
  String name;
  int age;

  Person(String name, int age) {
    this.name = name;
    this.age = age;
  }


  factory Person.fromJson(Map<String, dynamic> json) => _$PersonFromJson(json);

  Map<String, dynamic> toJson() => _$PersonToJson(this);
}



lib/src/dart_jsona_base.dart
import 'package:dart_jsona/src/builders/model_serializer.dart';
import 'package:dart_jsona/src/builders/property_mapper.dart';
import 'package:dart_jsona/src/builders/json_deserializer.dart';
import 'package:dart_jsona/src/models/jsona_types.dart';

class Jsona {
  AbsModelPropertiesMapper modelPropertiesMapper = new ModelPropertiesMapper();
  JsonPropertiesMapper jsonPropertiesMapper = new JsonPropertiesMapper();

  Map<String, Object> serialize({Map<String, Object> stuff, List<String> includeNames}) {
    if (stuff == null) {
      throw ArgumentError('Jsona can not serialize, stuff is not passed');
    }

    ModelSerializer modelSerializer = new ModelSerializer(modelPropertiesMapper);
    modelSerializer.setStuff(stuff);
    if (includeNames != null) {
      modelSerializer.setIncludedNames(includeNames);
    }

    return modelSerializer.build();
  }

  dynamic deserialize(Map<String, Object> body) {
    if (body == null) {
      throw ArgumentError('Json can not deserialize, body is not passed');
    }

    JsonDeserializer modelBuilder = new JsonDeserializer(jsonPropertiesMapper);
    modelBuilder.setJsonParsedObject(body);

    return modelBuilder.build();
  }
}


lib/src/models/jsona_types.dart
abstract class AbsModelPropertiesMapper {
  String getId(Map<String, dynamic> model);
  String getType(Map<String, dynamic> model);
  dynamic getAttributes(Map<String, dynamic> model);
  dynamic getRelationShips(Map<String, dynamic> model);
}

abstract class AbsJsonPropertiesMapper {
  Map<String, Object> createModel(String type);
  void setId(Map<String, Object> model, String id);
  void setAttributes(Map<String, Object> model, Map<String, Object> attributes);
  void setMeta(Map<String, Object> model, Map<String, Object>  meta);
  void setLinks(Map<String, Object>  model, Map<String, String> links);
  void setRelationships(Map<String, Object> model, Map<String, Object> relationships);
  void setRelationshipLinks(Map<String, Object> parentModel, String relationName, Map<String, String> links);
  void setRelationshipMeta(Map<String, Object> parentModel, String relationName, Map<String, Object> meta);
}

abstract class JsonaModelBuilder {
  Map<String, dynamic> build();
}


lib/src/builders/property_mapper.dart
import 'package:dart_jsona/src/models/jsona_types.dart';

final RELATIONSHIP_NAMES_PROP = 'relationshipNames';

class ModelPropertiesMapper implements AbsModelPropertiesMapper {
  @override
  String getId(Map<String, dynamic> model) {
    return model['id'];
  }

  @override
  String getType(Map<String, dynamic> model) {
    return model['type'];
  }

  @override
  dynamic getRelationShips(Map<String, dynamic> model) {
    List<String> relationshipNames = model[RELATIONSHIP_NAMES_PROP];

    if (relationshipNames == null) {
      return;
    }

    Map<String, dynamic> relationships = {};
    relationshipNames.forEach((relationName) {
      if (model[relationName] != null) {
        relationships[relationName] = model[relationName];
      }
    });

    return relationships;
  }

  @override
  dynamic getAttributes(Map<String, dynamic> model) {
    List<String> exceptProps = ['id', 'type', RELATIONSHIP_NAMES_PROP];

    if (model[RELATIONSHIP_NAMES_PROP] is List) {
      exceptProps.addAll(model[RELATIONSHIP_NAMES_PROP]);
    } else if (model[RELATIONSHIP_NAMES_PROP] != null) {
      print(
          "Can't getAttributes correctly, ${RELATIONSHIP_NAMES_PROP} property of ${model['type']}-${model['id']} model isn't array of relationship names");
    }

    var attributes = {};
    model.keys.forEach((attrName) {
      if (exceptProps.indexOf(attrName) == -1) {
        attributes[attrName] = model[attrName];
      }
    });

    return attributes;
  }
}

class JsonPropertiesMapper implements AbsJsonPropertiesMapper {
  @override
  Map<String, dynamic> createModel(String type) {
    return {'type': type};
  }

  @override
  void setId(Map<String, dynamic> model, String id) {
    model['id'] = id;
  }

  @override
  void setAttributes(Map<String, dynamic> model, Map<String, dynamic> attributes) {
    attributes.keys.forEach((propName) => model[propName] = attributes[propName]);
  }

  @override
  void setMeta(Map<String, dynamic> model, Map<String, dynamic> meta) {
    model['meta'] = meta;
  }

  @override
  void setLinks(Map<String, dynamic> model, Map<String, String> links) {
    model['links'] = links;
  }

  @override
  void setRelationshipLinks(Map<String, dynamic> parentModel, String relationName, Map<String, String> links) {
    // TODO: implement setRelationshipLinks
  }

  @override
  void setRelationshipMeta(Map<String, dynamic> parentModel, String relationName, Map<String, dynamic> meta) {
    // TODO: implement setRelationshipMeta
  }

  @override
  void setRelationships(Map<String, dynamic> model, Map<String, dynamic> relationships) {
    relationships.keys.forEach((propName) {
      model[propName] = relationships[propName];
    });

    List<String> newNames = relationships.keys.toList();

    model[RELATIONSHIP_NAMES_PROP] = newNames;
  }
}


lib/src/builders/model_serializer.dart
import 'package:dart_jsona/src/models/jsona_types.dart';
import 'package:dart_jsona/src/builders/util.dart';

class ModelSerializer implements JsonaModelBuilder {
  AbsModelPropertiesMapper propertiesMapper;
  Map<String, Object> stuff;
  Map<String, Object> includeNamesTree = {};

  ModelSerializer([AbsModelPropertiesMapper propertiesMapper]) {
    this.setPropertiesMapper(propertiesMapper);
  }

  void setPropertiesMapper(AbsModelPropertiesMapper setPropertiesMapper) {
    this.propertiesMapper = setPropertiesMapper;
  }

  void setStuff(Map<String, Object> stuff) {
    this.stuff = stuff;
  }

  void setIncludedNames(List<String> includeNames) {
    Map<String, Object> includeNamesTree = {};
    includeNames.forEach((nameChain) {
      createIncludeNamesTree(nameChain, includeNamesTree);
    });

    this.includeNamesTree = includeNamesTree;
  }

  @override
  Map<String, dynamic> build() {
    if (propertiesMapper == null) {
      throw ArgumentError('ModelsSerializer cannot build, propertiesMapper is not set');
    } else if (stuff == null) {
      throw ArgumentError('ModelsSerializer cannot build, stuff is not set');
    }

    Map<String, dynamic> body = {};
    Map<String, Object> uniqueIncluded = {};

    if (stuff != null && stuff is Iterable) {
      int collectionLength = stuff.length;
      List<Object> data = [];

      for (int i = 0; i < collectionLength; i++) {
        data.add(buildDataByModel(stuff[i]));

        buildIncludedByModel(stuff[i], this.includeNamesTree, uniqueIncluded);
      }
      body['data'] = data;
    } else if (stuff != null) {
      body['data'] = buildDataByModel(stuff);
      this.buildIncludedByModel(stuff, this.includeNamesTree, uniqueIncluded);
    } else if (stuff == null) {
      body['data'] = null;
    }

    if (uniqueIncluded.keys.length > 0) {
      body['included'] = [];
      List<String> includeUniqueKeys = uniqueIncluded.keys.toList();

      includeUniqueKeys.forEach((key) {
        body['included'].add(uniqueIncluded[key]);
      });
    }

    return body;
  }

  Map<String, Object> buildDataByModel(Map<String, Object> model) {
    Map<String, Object> data = {
      'id': propertiesMapper.getId(model),
      'type': propertiesMapper.getType(model),
      'attributes': propertiesMapper.getAttributes(model)
    };

    if (data['type'] == null) {
      throw 'ModelsSerializer cannot buildDataByModel, type is not set or incorrect';
    }

    var relationships = this.buildRelationshipsByModel(model);

    if (relationships != null) {
      data['relationships'] = relationships;
    }

    return data;
  }

  dynamic buildRelationshipsByModel(Map<String, dynamic> model) {
    var relations = propertiesMapper.getRelationShips(model);

    if (relations == null) {
      return;
    }

    var relationships = {};

    relations.keys.forEach((key) {
      dynamic relation = relations[key];

      if (relation is List) {
        List<dynamic> relationshipData = [];
        int relationLength = relation.length;

        for (int i = 0; i < relationLength; i++) {
          Map<String, Object> item = {
            'id': propertiesMapper.getId(relation[i]),
            'type': propertiesMapper.getType(relation[i])
          };

          if (item['id'] != null && item['type'] != null) {
            relationshipData.add(item);
          }
        }

        relationships[key] = {'data': relationshipData};
      } else if (relation != null) {
        Map<String, Object> item = {'id': propertiesMapper.getId(relation), 'type': propertiesMapper.getType(relation)};

        if (item['type'] != null) {
          relationships[key] = {'data': item};
        } else {
          print("Can't create data for relationship $key, it doesn't have 'type', it was skipped");
        }
      } else {
        relationships[key] = {'data': relation};
      }
    });

    return relationships;
  }

  void buildIncludedByModel(
      Map<String, Object> model, Map<String, dynamic> includeTree, Map<String, Object> builtIncluded) {
    if (includeTree == null) {
      return;
    }

    var modelRelationships = propertiesMapper.getRelationShips(model);

    if (modelRelationships == null) {
      return;
    }

    var includeNames = includeTree.keys.toList();
    var includeNamesLength = includeNames.length;

    for (int i = 0; i < includeNamesLength; i++) {
      String currentRelationName = includeNames[i];
      var relation = modelRelationships[currentRelationName];

      if (relation != null) {
        if (relation is List) {
          var relationModelsLength = relation.length;

          for (int r = 0; r < relationModelsLength; r++) {
            var relationModel = relation[r];
            buildIncludedItem(relationModel, includeTree[currentRelationName], builtIncluded);
          }
        } else {
          buildIncludedItem(relation, includeTree[currentRelationName], builtIncluded);
        }
      }
    }
  }

  void buildIncludedItem(Map<String, dynamic> relationModel, subIncludeTree, builtIncluded) {
    var includeKey = propertiesMapper.getType(relationModel) + propertiesMapper.getId(relationModel);

    if (builtIncluded[includeKey] == null) {
      builtIncluded[includeKey] = buildDataByModel(relationModel);

      if (subIncludeTree != null) {
        this.buildIncludedByModel(relationModel, subIncludeTree, builtIncluded);
      }
    }
  }
}


lib/src/builders/json_deserializer.dart
import 'package:dart_jsona/src/models/jsona_types.dart';

class JsonDeserializer {
  dynamic body;
  AbsJsonPropertiesMapper propertiesMapper;
  Map<String, Object> includedInObject;
  Map<String, Object> cachedModels = {};

  JsonDeserializer(propertiesMapper) {
    this.setPropertiesMapper(propertiesMapper);
  }

  void setPropertiesMapper(pm) {
    this.propertiesMapper = pm;
  }

  String createEntityKey(data) {
    if (data['type'] != null && data['id'] != null) {
      return '${data['type']}-${data['id']}';
    }
    return '';
  }

  void setJsonParsedObject(Map<String, dynamic> body) {
    this.body = body;
  }

  dynamic build() {
    var data = body['data'];
    var stuff;

    if (data is List) {
      stuff = [];
      var collectionLength = data.length;

      for (var i = 0; i < collectionLength; i++) {
        if (data[i] != null) {
          var model = buildModelByData(data[i]);

          if (model != null) {
            stuff.add(model);
          }
        }
      }
    } else if (data != null) {
      stuff = buildModelByData(data);
    }

    return stuff;
  }

  Map<String, dynamic> buildModelByData(Map<String, Object> data) {
    var entityKey = createEntityKey(data);

    var model = propertiesMapper.createModel(data['type']);

    if (model != null) {
      if (entityKey != null) {
        cachedModels[entityKey] = model;
      }
      propertiesMapper.setId(model, data['id']);

      if (data['attributes'] != null) {
        propertiesMapper.setAttributes(model, data['attributes']);
      }

      if (data['meta'] != null) {
        this.propertiesMapper.setMeta(model, data['meta']);
      }

      if (data['links'] != null) {
        this.propertiesMapper.setLinks(model, data['links']);
      }

      var relationships = buildRelationsByData(data, model);

      if (relationships != null) {
        propertiesMapper.setRelationships(model, relationships);
      }
    }

    return model;
  }

  Map<String, dynamic> buildRelationsByData(Map<String, dynamic> data, Map<String, dynamic> model) {
    Map<String, dynamic> readyRelations = {};

    if (data['relationships'] != null) {
      if (data['relationships'] is Map) {
        data['relationships'].keys.forEach((k) {
          Map<String, dynamic> relation = data['relationships'][k];

          if (relation['data'] is Iterable) {
            readyRelations[k] = [];

            int relationItemsLength = relation['data'].length;
            var relationItem;

            for (int i = 0; i < relationItemsLength; i++) {
              relationItem = relation['data'][i];

              if (relationItem == null) {
                return null;
              }

              var dataItem = buildDataFromIncludedOrData(relationItem['id'], relationItem['type']);
              readyRelations[k].add(this.buildModelByData(dataItem));
            }
          } else if (relation['data'] != null) {
            var dataItem = buildDataFromIncludedOrData(relation['data']['id'], relation['data']['type']);
            readyRelations[k] = buildModelByData(dataItem);
          } else if (relation['data'] == null) {
            readyRelations[k] = null;
          }

          if (relation['links'] != null) {
            this.propertiesMapper.setRelationshipLinks(model, k, relation['links']);
          }

          if (relation['meta'] != null) {
            this.propertiesMapper.setRelationshipMeta(model, k, relation['meta']);
          }
        });
      }
    }

    if (readyRelations.keys.length > 0) {
      return readyRelations;
    }

    return null;
  }

  Map<String, Object> buildDataFromIncludedOrData(String id, String type) {
    Map<String, Object> included = buildIncludedInObject();
    var dataItem = included[type + id];

    if (dataItem != null) {
      return dataItem;
    } else {
      return {'id': id, 'type': type};
    }
  }

  Map<String, Object> buildIncludedInObject() {
    if (includedInObject == null) {
      this.includedInObject = {};

      if (body['included'] != null) {
        int includedLength = body['included'].length;

        for (int i = 0; i < includedLength; i++) {
          var item = body['included'][i];
          var key = item['type'] + item['id'];
          includedInObject[key] = item;
        }
      }
    }

    return includedInObject;
  }
}


example/dart_jsona_example.dart
import 'package:dart_jsona/dart_jsona.dart';

void main() {
  Jsona jsona = new Jsona();

  Map<String, dynamic> body = {
    'data': {
      'type': 'town',
      'id': '123',
      'attributes': {
        'name': 'Barcelona',
      },
      'relationships': {
        'country': {
          'data': {
            'type': 'country',
            'id': '32',
          }
        }
      }
    },
    'included': [{
      'type': 'country',
      'id': '32',
      'attributes': {
        'name': 'Spain',
      }
    }]
  };

  final dynamic town = jsona.deserialize(body);

  print(town);
  /**
   * {
   *  type: town,
   *  id: 123,
   *  name: Barcelona,
   *  country: {
   *    type: country,
   *    id: 32, name: Spain
   *  },
   *  relationshipNames: [country]}
   */


  Map<String, Object> jsonBody = jsona.serialize(stuff: town, includeNames: ['country']);

  print(jsonBody);
  /**
   * {
   *   data: {
   *     id: 123,
   *     type: town,
   *     attributes: {
   *       name: Barcelona
   *     },
   *     relationships: {
   *       country: {
   *         data: {
   *           id: 32, type: country
   *         }
   *       }
   *     }
   *   },
   *   included: [{
   *     id: 32,
   *     type: country,
   *     attributes: {
   *       name: Spain
   *     }
   *   }]
   * }
   */
}
